src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -0,0 +1,403 @@
+using size_t = System.IntPtr;
+using FSEventStreamEventId = System.UInt64;
+using CFTimeInterval = System.Double;
+using CFRunLoopRef = System.IntPtr;
+
+namespace System.IO
+{
+    // OSX implementation of FileSystemWatcher's PAL.  The main partial definition of FileSystemWatcher expects:
+    // - StartRaisingEvents
+    // - StopRaisingEvents
+    // - FinalizeDispose
+
+    public partial class FileSystemWatcher
+    {
+        // OS X has a case-sensitive File System
+        internal const bool CaseSensitive = true;
+
+        // Flags used to create the event stream
+        private const Interop.CoreServices.FSEventStreamCreateFlags EventStreamFlags = (Interop.CoreServices.FSEventStreamCreateFlags.kFSEventStreamCreateFlagFileEvents |
+                                                                                        Interop.CoreServices.FSEventStreamCreateFlags.kFSEventStreamCreateFlagWatchRoot);
+
+        // The last directory we were told to watch. This allows us to be smart about how we create streams
+        // and to only teardown and recreate streams when we need to.
+        private String _lastWatchedDirectory = String.Empty;
+
+        // The bitmask of events that we want to send to the user
+        private Interop.CoreServices.FSEventStreamEventFlags _filterFlags;
+
+        // The EventStream to listen for events on
+        private FSEventStreamRef _eventStream = IntPtr.Zero;
+
+        // The background thread to use to monitor events
+        private Thread _watcherThread = null;
+
+        // A reference to the RunLoop that we can use to start or stop a Watcher
+        private CFRunLoopRef _watcherRunLoop = IntPtr.Zero;
+
+        // Use an event to try to prevent StartRaisingEvents from returning before the
+        // RunLoop actually begins. This will mitigate a race condition where the watcher
+        // thread hasn't completed initialization and stop is called before the RunLoop even starts.
+        private AutoResetEvent _runLoopStartedEvent = new AutoResetEvent(false);
+
+        /// <summary>Called when FileSystemWatcher is finalized.</summary>
+        private void FinalizeDispose()
+        {
+            // Make sure we cleanup
+            StopRaisingEvents();
+            CleanupStreamAndWatcher();
+        }
+
+        private void StartRaisingEvents()
+        {
+            // Make sure the Start and Stop can be called from different threads and don't 
+            // stomp on the other's operation. We use the _lastWatchedDirectory instead of the
+            // RunLoop or StreamRef because IntPtrs are value types and can't be locked
+            lock (_lastWatchedDirectory)
+            {
+                // Always re-create the filter flags when start is called since they could have changed
+                if (_notifyFilters.HasFlag(NotifyFilters.Attributes | NotifyFilters.CreationTime | NotifyFilters.LastAccess | NotifyFilters.LastWrite | NotifyFilters.Size))
```

Comment:
Do you need to initially set _filterFlags to 0 here?  It seems like if this initial if condition is false, you may end up with the previous value _filterFlags was set to, and then the subsequent if blocks or in additional values.